cli/src/commands/sign.rs

```diff
@@ -0,0 +1,134 @@
+use crate::cli_util::RevisionArg;
+use crate::command_error::user_error_with_hint;
+use crate::command_error::CommandError;
+use crate::complete;
+use crate::ui::Ui;
+
+/// Cryptographically sign a revision
+#[derive(clap::Args, Clone, Debug)]
+pub struct SignArgs {
+    /// What key to use, depends on the configured signing backend.
+    #[arg()]
+    key: Option<String>,
+    /// What revision(s) to sign
+    ///
+    /// Note that revisions, which are already signed by you, will be resigned.
+    ///
+    /// While that leads to discomfort for users, which sign with hardware
+    /// devices, as of now we cannot reliably check if a commit is already
+    /// signed by the user without creating a signature (see https://github.com/jj-vcs/jj/issues/5786).
+    #[arg(
+        long, short,
+        value_name = "REVSETS",
+        add = ArgValueCandidates::new(complete::mutable_revisions),
+    )]
+    revisions: Vec<RevisionArg>,
+}
+
+pub fn cmd_sign(ui: &mut Ui, command: &CommandHelper, args: &SignArgs) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+
+    if !workspace_command.repo().store().signer().can_sign() {
+        return Err(user_error_with_hint(
+            "No signing backend configured",
+            "For configuring a signing backend, see https://jj-vcs.github.io/jj/latest/config/#commit-signing",
+        ));
+    }
+
+    let to_sign: IndexSet<Commit> = workspace_command
+        .parse_union_revsets(ui, &args.revisions)?
+        .evaluate_to_commits()?
+        .try_collect()?;
+
+    workspace_command.check_rewritable(to_sign.iter().ids())?;
+
+    let mut tx = workspace_command.start_transaction();
+
+    let mut signed_commits = vec![];
+    let mut num_reparented = 0;
+
+    tx.repo_mut().transform_descendants(
+        to_sign.iter().ids().cloned().collect_vec(),
+        |rewriter| {
+            let old_commit = rewriter.old_commit().clone();
+            let commit_builder = rewriter.reparent();
+
+            if to_sign.contains(&old_commit) {
+                let new_commit = commit_builder
+                    .set_sign_key(args.key.clone())
+                    .set_sign_behavior(SignBehavior::Force)
+                    .write()?;
+
+                signed_commits.push(new_commit);
+            } else {
+                commit_builder.write()?;
+                num_reparented += 1;
+            }
+
+            Ok(())
+        },
+    )?;
+
+    if let Some(mut formatter) = ui.status_formatter() {
+        if !signed_commits.is_empty() {
+            let template = tx.commit_summary_template();
+            writeln!(formatter, "Signed {} commit(s):", signed_commits.len())?;
+            for commit in &signed_commits {
+                write!(formatter, "  ")?;
+                template.format(commit, formatter.as_mut())?;
+                writeln!(formatter)?;
+            }
+        }
+    }
+
+    let num_not_authored_by_me = to_sign
```

Comment:
nit: I think `signed_commits` is more correct here.

Reply:
Oh yeah, absolutely. Thanks for spotting!